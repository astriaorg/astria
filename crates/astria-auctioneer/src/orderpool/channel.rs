use jiff::Timestamp;
use tokio::sync::mpsc;
use uuid::Uuid;

use super::{
    in_memory::{
        InsertedOrReplaced,
        RemovedOrNotFound,
    },
    Order,
};

pub(crate) struct Request {
    pub(super) order: Order,
    pub(super) to_requester: tokio::sync::oneshot::Sender<Response>,
}

pub(crate) struct ForOrder {
    pub(crate) uuid: Uuid,
    pub(crate) action: InsertedOrReplaced,
}
pub(crate) struct ForCancellation {
    pub(crate) uuid: Uuid,
    pub(crate) timestamp: Timestamp,
    pub(crate) action: RemovedOrNotFound,
}

pub(crate) enum Response {
    ForOrder(ForOrder),
    ForCancellation(ForCancellation),
}

impl From<ForCancellation> for Response {
    fn from(value: ForCancellation) -> Self {
        Self::ForCancellation(value)
    }
}
impl From<ForOrder> for Response {
    fn from(value: ForOrder) -> Self {
        Self::ForOrder(value)
    }
}
pub(super) fn new() -> (Sender, Receiver) {
    let (tx, rx) = mpsc::channel(32);
    (
        Sender {
            inner: tx,
        },
        Receiver {
            inner: rx,
        },
    )
}

pub(super) struct Receiver {
    inner: mpsc::Receiver<Request>,
}

impl Receiver {
    pub(super) async fn recv(&mut self) -> Option<Request> {
        self.inner.recv().await
    }
}

#[derive(Debug, thiserror::Error)]
pub(crate) enum SendError {
    #[error("no available capacity in order pool")]
    Full,
    #[error("order pool channel is closed")]
    Closed,
    #[error("order pool dropped the response channel before sending a response")]
    Dropped,
}

impl From<tokio::sync::mpsc::error::TrySendError<Request>> for SendError {
    fn from(value: tokio::sync::mpsc::error::TrySendError<Request>) -> Self {
        match value {
            mpsc::error::TrySendError::Full(_) => Self::Full,
            mpsc::error::TrySendError::Closed(_) => Self::Closed,
        }
    }
}

impl From<tokio::sync::oneshot::error::RecvError> for SendError {
    fn from(_value: tokio::sync::oneshot::error::RecvError) -> Self {
        Self::Dropped
    }
}

#[derive(Clone)]
pub(crate) struct Sender {
    inner: mpsc::Sender<Request>,
}

impl Sender {
    pub(crate) async fn send(&self, order: Order) -> Result<Response, SendError> {
        let (tx, rx) = tokio::sync::oneshot::channel();
        self.inner.try_send(Request {
            order,
            to_requester: tx,
        })?;
        let rsp = rx.await?;
        Ok(rsp)
    }
}
