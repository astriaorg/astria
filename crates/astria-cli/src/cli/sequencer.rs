use std::str::FromStr;

use astria_sequencer_client::Address;
use clap::{
    Args,
    Subcommand,
};
use color_eyre::{
    eyre,
    eyre::Context,
};

/// Interact with a Sequencer node
#[derive(Debug, Subcommand)]
pub enum Command {
    /// Commands for interacting with Sequencer accounts
    Account {
        #[clap(subcommand)]
        command: AccountCommand,
    },
    /// Commands for interacting with Sequencer balances
    Balance {
        #[clap(subcommand)]
        command: BalanceCommand,
    },
    /// Commands for interacting with Sequencer block heights
    #[clap(name = "blockheight")]
    BlockHeight {
        #[clap(subcommand)]
        command: BlockHeightCommand,
    },
}

#[derive(Debug, Subcommand)]
pub enum AccountCommand {
    /// Create a new Sequencer account
    Create,
}

#[derive(Debug, Subcommand)]
pub enum BalanceCommand {
    /// Get the balance of a Sequencer account
    Get(BalanceGetArgs),
}

#[derive(Args, Debug)]
pub struct BalanceGetArgs {
    /// The url of the Sequencer node
    #[clap(long)]
    pub(crate) sequencer_url: String,
    /// The address of the Sequencer account
    pub(crate) address: SequencerAddressArg,
}

#[derive(Clone, Debug, PartialEq)]
pub(crate) struct SequencerAddressArg(pub(crate) Address);

impl FromStr for SequencerAddressArg {
    type Err = eyre::Report;

    /// Parse a string into a Sequencer Address
    fn from_str(s: &str) -> eyre::Result<Self, Self::Err> {
        let address_bytes = hex::decode(s).wrap_err(
            "failed to decode address. address should be 20 bytes long. do not prefix with 0x",
        )?;
        let address =
            Address::try_from_slice(address_bytes.as_ref()).wrap_err("failed to create address")?;

        Ok(Self(address))
    }
}

#[derive(Debug, Subcommand)]
pub enum BlockHeightCommand {
    /// Get the current block height of the Sequencer node
    Get(BlockHeightGetArgs),
}

#[derive(Args, Debug)]
pub struct BlockHeightGetArgs {
    /// The url of the Sequencer node
    #[clap(long)]
    pub(crate) sequencer_url: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sequencer_address_arg_from_str_valid() {
        let hex_str = "a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0";
        let bytes = hex::decode(hex_str).unwrap();
        let expected_address = Address::try_from_slice(&bytes).unwrap();

        let sequencer_address_arg: SequencerAddressArg = hex_str.parse().unwrap();
        assert_eq!(sequencer_address_arg, SequencerAddressArg(expected_address));
    }

    #[test]
    fn test_sequencer_address_arg_from_str_invalid() {
        let hex_str = "invalidhexstr";
        let result: eyre::Result<SequencerAddressArg> = hex_str.parse();
        assert!(result.is_err());

        let error_message = format!("{:?}", result.unwrap_err());
        assert!(error_message.contains("failed to decode address"));
    }
}
