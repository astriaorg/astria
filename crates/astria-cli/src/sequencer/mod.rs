use clap::Subcommand;
use color_eyre::eyre;

mod account;
mod address;
mod balance;
mod block_height;
mod bridge_lock;
mod ics20_withdrawal;
mod init_bridge_account;
mod submit;
mod sudo;
mod transfer;
mod sign;

#[derive(Debug, clap::Args)]
pub(super) struct Command {
    #[command(subcommand)]
    command: SubCommand,
}

impl Command {
    pub(super) async fn run(self) -> eyre::Result<()> {
        match self.command {
            SubCommand::Account(account) => account.run().await,
            SubCommand::Address(address) => address.run(),
            SubCommand::Balance(balance) => balance.run().await,
            SubCommand::BlockHeight(block_height) => block_height.run().await,
            SubCommand::BridgeLock(bridge_lock) => bridge_lock.run().await,
            SubCommand::InitBridgeAccount(init_bridge_account) => init_bridge_account.run().await,
            SubCommand::Sudo(sudo) => sudo.run().await,
            SubCommand::Transfer(transfer) => transfer.run().await,
            SubCommand::Ics20Withdrawal(ics20_withdrawal) => ics20_withdrawal.run().await,
            SubCommand::Submit(submit) => submit.run().await,
            SubCommand::Sign(sign) => sign.run().await,

        }
    }
}

/// Interact with a Sequencer node
#[derive(Debug, Subcommand)]
enum SubCommand {
    /// Commands for interacting with Sequencer accounts
    Account(account::Command),
    /// Utilities for constructing and inspecting sequencer addresses
    Address(address::Command),
    /// Commands for interacting with Sequencer balances
    Balance(balance::Command),
    /// Commands for interacting with Sequencer block heights
    #[command(name = "blockheight")]
    BlockHeight(block_height::Command),
    /// Command for transferring to a bridge account
    BridgeLock(bridge_lock::Command),
    /// Command for initializing a bridge account
    InitBridgeAccount(init_bridge_account::Command),
    /// Commands requiring authority for Sequencer
    Sudo(sudo::Command),
    /// Command for sending balance between accounts
    Transfer(transfer::Command),
    /// Command for withdrawing an ICS20 asset
    Ics20Withdrawal(ics20_withdrawal::Command),
    /// Command for submitting a presigned message
    Submit(submit::Command),
    Sign(sign::Command),
}
