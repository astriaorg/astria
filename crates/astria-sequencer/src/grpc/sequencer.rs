use std::sync::Arc;

use astria_core::{
    generated::sequencerblock::v1alpha1::{
        sequencer_service_server::SequencerService,
        FilteredSequencerBlock as RawFilteredSequencerBlock,
        GetFilteredSequencerBlockRequest,
        GetSequencerBlockRequest,
        SequencerBlock as RawSequencerBlock,
    },
    primitive::v1::RollupId,
};
use cnidarium::Storage;
use tonic::{
    Request,
    Response,
    Status,
};
use tracing::instrument;

use crate::{
    api_state_ext::StateReadExt as _,
    state_ext::StateReadExt as _,
};

pub(crate) struct SequencerServer {
    storage: Storage,
}

impl SequencerServer {
    pub(crate) fn new(storage: Storage) -> Self {
        Self {
            storage,
        }
    }
}

#[async_trait::async_trait]
impl SequencerService for SequencerServer {
    /// Given a block height, returns the sequencer block at that height.
    #[instrument(skip_all, fields(height = request.get_ref().height))]
    async fn get_sequencer_block(
        self: Arc<Self>,
        request: Request<GetSequencerBlockRequest>,
    ) -> Result<Response<RawSequencerBlock>, Status> {
        let snapshot = self.storage.latest_snapshot();
        let curr_block_height = snapshot.get_block_height().await.map_err(|e| {
            Status::internal(format!("failed to get block height from storage: {e}"))
        })?;

        let request = request.into_inner();

        if curr_block_height < request.height {
            return Err(Status::invalid_argument(
                "requested height is greater than current block height",
            ));
        }

        let block = snapshot
            .get_sequencer_block_by_height(request.height)
            .await
            .map_err(|e| {
                Status::internal(format!("failed to get sequencer block from storage: {e}"))
            })?;

        Ok(Response::new(block.into_raw()))
    }

    /// Given a block height and set of rollup ids, returns a SequencerBlock which
    /// is filtered to contain only the transactions that are relevant to the given rollup.
    #[instrument(skip_all, fields(height = request.get_ref().height))]
    async fn get_filtered_sequencer_block(
        self: Arc<Self>,
        request: Request<GetFilteredSequencerBlockRequest>,
    ) -> Result<Response<RawFilteredSequencerBlock>, Status> {
        let snapshot = self.storage.latest_snapshot();
        let curr_block_height = snapshot.get_block_height().await.map_err(|e| {
            Status::internal(format!("failed to get block height from storage: {e}"))
        })?;

        let request = request.into_inner();

        if curr_block_height < request.height {
            return Err(Status::invalid_argument(
                "requested height is greater than current block height",
            ));
        }

        let rollup_ids = request
            .rollup_ids
            .iter()
            .map(RollupId::try_from_raw)
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| Status::invalid_argument(format!("invalid rollup ID: {e}")))?;

        let block_hash = snapshot
            .get_block_hash_by_height(request.height)
            .await
            .map_err(|e| Status::internal(format!("failed to get block hash from storage: {e}")))?;

        let header = snapshot
            .get_sequencer_block_header_by_hash(&block_hash)
            .await
            .map_err(|e| {
                Status::internal(format!(
                    "failed to get sequencer block header from storage: {e}"
                ))
            })?;

        let (rollup_transactions_proof, rollup_ids_proof) = snapshot
            .get_block_proofs_by_block_hash(&block_hash)
            .await
            .map_err(|e| {
                Status::internal(format!(
                    "failed to get sequencer block proofs from storage: {e}"
                ))
            })?;

        let mut all_rollup_ids = snapshot
            .get_rollup_ids_by_block_hash(&block_hash)
            .await
            .map_err(|e| Status::internal(format!("failed to get rollup ids from storage: {e}")))?;
        all_rollup_ids.sort_unstable();

        // Filter out the Rollup Ids requested which have no data before grabbing
        // so as to not error because the block had no data for the requested rollup
        let rollup_ids: Vec<RollupId> = rollup_ids
            .into_iter()
            .filter(|id| all_rollup_ids.binary_search(id).is_ok())
            .collect();
        let mut rollup_transactions = Vec::with_capacity(rollup_ids.len());
        for rollup_id in rollup_ids {
            let rollup_data = snapshot
                .get_rollup_data(&block_hash, &rollup_id)
                .await
                .map_err(|e| {
                    Status::internal(format!("failed to get rollup data from storage: {e}",))
                })?;
            rollup_transactions.push(rollup_data.into_raw());
        }

        let all_rollup_ids = all_rollup_ids.into_iter().map(RollupId::to_vec).collect();

        let block = RawFilteredSequencerBlock {
            block_hash: block_hash.to_vec(),
            header: Some(header.into_raw()),
            rollup_transactions,
            rollup_transactions_proof: rollup_transactions_proof.into(),
            rollup_ids_proof: rollup_ids_proof.into(),
            all_rollup_ids,
        };

        Ok(Response::new(block))
    }
}

#[cfg(test)]
mod test {
    use astria_core::{
        protocol::test_utils::ConfigureSequencerBlock,
        sequencerblock::v1alpha1::SequencerBlock,
    };
    use cnidarium::StateDelta;

    use super::*;
    use crate::{
        api_state_ext::StateWriteExt as _,
        state_ext::StateWriteExt,
    };

    fn make_test_sequencer_block(height: u32) -> SequencerBlock {
        ConfigureSequencerBlock {
            height,
            ..Default::default()
        }
        .make()
    }

    #[tokio::test]
    async fn test_get_sequencer_block() {
        let block = make_test_sequencer_block(1);
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let mut state_tx = StateDelta::new(storage.latest_snapshot());
        state_tx.put_block_height(1);
        state_tx.put_sequencer_block(block.clone()).unwrap();
        storage.commit(state_tx).await.unwrap();

        let server = Arc::new(SequencerServer::new(storage.clone()));
        let request = GetSequencerBlockRequest {
            height: 1,
        };
        let request = Request::new(request);
        let response = server.get_sequencer_block(request).await.unwrap();
        assert_eq!(response.into_inner().header.unwrap().height, 1);
    }
}
