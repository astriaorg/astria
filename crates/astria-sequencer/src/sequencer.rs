use anyhow::{
    anyhow,
    Context as _,
    Result,
};
use penumbra_tower_trace::{
    trace::request_span,
    RequestExt as _,
};
use tendermint::abci::ConsensusRequest;
use tower_abci::v037::Server;
use tracing::{
    info,
    instrument,
};

use crate::{
    app::App,
    config::Config,
    service,
};

pub struct Sequencer;

impl Sequencer {
    #[instrument(skip_all)]
    pub async fn run_until_stopped(config: Config) -> Result<()> {
        if config
            .db_filepath
            .try_exists()
            .context("failed checking for existence of db storage file")?
        {
            info!(
                path = %config.db_filepath.display(),
                "opening storage db"
            );
        } else {
            info!(
                path = %config.db_filepath.display(),
                "creating storage db"
            );
        }
        let storage = penumbra_storage::Storage::load(config.db_filepath.clone())
            .await
            .context("failed to load storage backing chain state")?;
        let snapshot = storage.latest_snapshot();
        let app = App::new(snapshot);

        let consensus_service = tower::ServiceBuilder::new()
            .layer(request_span::layer(|req: &ConsensusRequest| {
                req.create_span()
            }))
            .service(tower_actor::Actor::new(10, |queue: _| {
                let storage = storage.clone();
                async move { service::Consensus::new(storage, app, queue).run().await }
            }));
        let mempool_service = service::Mempool;
        let info_service =
            service::Info::new(storage.clone()).context("failed initializing info service")?;
        let snapshot_service = service::Snapshot;

        let server = Server::builder()
            .consensus(consensus_service)
            .info(info_service)
            .mempool(mempool_service)
            .snapshot(snapshot_service)
            .finish()
            .ok_or_else(|| anyhow!("server builder didn't return server; are all fields set?"))?;

        info!(config.listen_addr, "starting sequencer");
        server
            .listen(&config.listen_addr)
            .await
            .expect("should listen");
        Ok(())
    }
}
