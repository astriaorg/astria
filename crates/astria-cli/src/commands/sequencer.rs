use astria_core::sequencer::v1alpha1::{
    transaction::action::{
        Action,
        TransferAction,
    },
    UnsignedTransaction,
};
use astria_sequencer_client::{
    Address,
    HttpClient,
    SequencerClientExt,
};
use color_eyre::{
    eyre,
    eyre::{
        ensure,
        eyre,
        Context,
    },
};
use ed25519_consensus::SigningKey;
use rand::rngs::OsRng;

use crate::cli::sequencer::{
    BasicAccountArgs,
    BlockHeightGetArgs,
    TransferArgs,
};

/// Generate a new signing key (this is also called a secret key by other implementations)
fn get_new_signing_key() -> SigningKey {
    SigningKey::new(OsRng)
}

/// Get the public key from the signing key
fn get_public_key_pretty(signing_key: &SigningKey) -> String {
    let verifying_key_bytes = signing_key.verification_key().to_bytes();
    hex::encode(verifying_key_bytes)
}

/// Get the private key from the signing key
fn get_private_key_pretty(signing_key: &SigningKey) -> String {
    let secret_key_bytes = signing_key.to_bytes();
    hex::encode(secret_key_bytes)
}

/// Get the address from the signing key
fn get_address_pretty(signing_key: &SigningKey) -> String {
    let address = Address::from_verification_key(signing_key.verification_key());
    hex::encode(address.to_vec())
}

/// Generates a new ED25519 keypair and prints the public key, private key, and address
pub(crate) fn create_account() {
    let signing_key = get_new_signing_key();
    let public_key_pretty = get_public_key_pretty(&signing_key);
    let private_key_pretty = get_private_key_pretty(&signing_key);
    let address_pretty = get_address_pretty(&signing_key);

    println!("Create Sequencer Account");
    println!();
    // TODO: don't print private keys to CLI, prefer writing to file:
    // https://github.com/astriaorg/astria/issues/594
    println!("Private Key: {private_key_pretty:?}");
    println!("Public Key:  {public_key_pretty:?}");
    println!("Address:     {address_pretty:?}");
}

/// Gets the balance of a Sequencer account
///
/// # Arguments
///
/// * `args` - The arguments passed to the command
///
/// # Errors
///
/// * If the http client cannot be created
/// * If the balance cannot be retrieved
pub(crate) async fn get_balance(args: &BasicAccountArgs) -> eyre::Result<()> {
    let address = &args.address;
    let sequencer_client = HttpClient::new(args.sequencer_url.as_str())
        .wrap_err("failed constructing http sequencer client")?;

    let res = sequencer_client
        .get_latest_balance(address.0)
        .await
        .wrap_err("failed to get balance")?;

    println!("Balances for address {}:", address.0);
    for balance in res.balances {
        println!("    asset ID: {}", hex::encode(balance.denom.id()));
        println!("    {} {}", balance.balance, balance.denom);
    }

    Ok(())
}

// Gets the balance of a Sequencer account
/// # Arguments
///
/// * `args` - The arguments passed to the command
///
/// # Errors
///
/// * If the http client cannot be created
/// * If the balance cannot be retrieved
pub(crate) async fn get_nonce(args: &BasicAccountArgs) -> eyre::Result<()> {
    let address = &args.address;
    let sequencer_client = HttpClient::new(args.sequencer_url.as_str())
        .wrap_err("failed constructing http sequencer client")?;

    let res = sequencer_client
        .get_latest_nonce(address.0)
        .await
        .wrap_err("failed to get nonce")?;

    println!("Nonce for address {}:", address.0);
    println!("    {} at height {}", res.nonce, res.height);

    Ok(())
}

/// Gets the latest block height of a Sequencer node
///
/// # Arguments
///
/// * `args` - The arguments passed to the command
///
/// # Errors
///
/// * If the http client cannot be created
/// * If the latest block height cannot be retrieved
pub(crate) async fn get_block_height(args: &BlockHeightGetArgs) -> eyre::Result<()> {
    let sequencer_client = HttpClient::new(args.sequencer_url.as_str())
        .wrap_err("failed constructing http sequencer client")?;

    let res = sequencer_client
        .latest_sequencer_block()
        .await
        .wrap_err("failed to get sequencer block")?;

    println!("Block Height:");
    println!("    {}", res.header().height);

    Ok(())
}

/// Gets the latest block height of a Sequencer node
///
/// # Arguments
///
/// * `args` - The arguments passed to the command
///
/// # Errors
///
/// * If the http client cannot be created
/// * If the latest block height cannot be retrieved
pub(crate) async fn send_transfer(args: &TransferArgs) -> eyre::Result<()> {
    // Build the signing_key
    let private_key_bytes: [u8; 32] = hex::decode(args.private_key.as_str())
        .wrap_err("failed to decode private key bytes from hex string")?
        .try_into()
        .map_err(|_| eyre!("invalid private key length; must be 32 bytes"))?;
    let sequencer_key =
        SigningKey::try_from(private_key_bytes).wrap_err("failed to parse sequencer key")?;

    // To and from addresses
    let from_address = Address::from_verification_key(sequencer_key.verification_key());
    let to_address = args.to_address.0;

    let sequencer_client = HttpClient::new(args.sequencer_url.as_str())
        .wrap_err("failed constructing http sequencer client")?;

    // Fetch the nonce for the action
    let nonce_res = sequencer_client
        .get_latest_nonce(from_address)
        .await
        .wrap_err("failed to get nonce")?;

    // Build and submit tx
    let tx = UnsignedTransaction {
        nonce: nonce_res.nonce,
        actions: vec![Action::Transfer(TransferAction {
            to: to_address,
            amount: args.amount,
            asset_id: astria_core::sequencer::v1alpha1::asset::default_native_asset_id(),
        })],
        fee_asset_id: astria_core::sequencer::v1alpha1::asset::default_native_asset_id(),
    }
    .into_signed(&sequencer_key);
    let res = sequencer_client
        .submit_transaction_commit(tx)
        .await
        .wrap_err("failed to submit transfer transaction")?;

    ensure!(res.tx_result.code.is_ok(), "error with transfer");
    println!("Transfer completed!");

    Ok(())
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_get_new_signing_key() {
        // generates seed of 32 bytes
        let key1 = get_new_signing_key();
        assert_eq!(key1.to_bytes().len(), 32, "Signing key is not 32 bytes");

        // generates different values
        let key2 = get_new_signing_key();
        assert_ne!(
            key1.to_bytes(),
            key2.to_bytes(),
            "Two signing key seeds are unexpectedly equal"
        );
    }

    #[test]
    fn test_signing_key_is_valid() {
        let key = get_new_signing_key();
        let msg = "Hello, world!";
        let signature = key.sign(msg.as_bytes());

        let verification_key = key.verification_key();
        assert!(
            verification_key.verify(&signature, msg.as_bytes()).is_ok(),
            "Signature verification failed"
        );
    }

    #[test]
    fn test_get_public_key_pretty() {
        let signing_key = get_new_signing_key();
        let public_key_pretty = get_public_key_pretty(&signing_key);
        assert_eq!(public_key_pretty.len(), 64);
    }

    #[test]
    fn test_get_private_key_pretty() {
        let signing_key = get_new_signing_key();
        let private_key_pretty = get_private_key_pretty(&signing_key);
        assert_eq!(private_key_pretty.len(), 64);
    }

    #[test]
    fn test_get_address_pretty() {
        let signing_key = get_new_signing_key();
        let address_pretty = get_address_pretty(&signing_key);
        assert_eq!(address_pretty.len(), 40);
    }
}
