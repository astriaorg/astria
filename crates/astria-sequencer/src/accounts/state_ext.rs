use anyhow::{
    Context,
    Result,
};
use async_trait::async_trait;
use borsh::{
    BorshDeserialize,
    BorshSerialize,
};
use hex::ToHex as _;
use ibc_types::core::channel::ChannelId;
use penumbra_storage::{
    StateRead,
    StateWrite,
};
use proto::native::sequencer::v1alpha1::{
    asset,
    Address,
};
use tracing::{
    debug,
    instrument,
};

/// Newtype wrapper to read and write a u32 from rocksdb.
#[derive(BorshSerialize, BorshDeserialize, Debug)]
struct Nonce(u32);

/// Newtype wrapper to read and write a u128 from rocksdb.
#[derive(BorshSerialize, BorshDeserialize, Debug)]
struct Balance(u128);

const ACCOUNTS_PREFIX: &str = "accounts";

fn storage_key(address: &str) -> String {
    format!("{ACCOUNTS_PREFIX}/{address}")
}

fn balance_storage_key(address: Address, asset: asset::Id) -> String {
    format!(
        "{}/balance/{}",
        storage_key(&address.encode_hex::<String>()),
        asset.encode_hex::<String>()
    )
}

fn nonce_storage_key(address: Address) -> String {
    format!("{}/nonce", storage_key(&address.encode_hex::<String>()))
}

fn channel_balance_storage_key(channel: &ChannelId, asset: asset::Id) -> String {
    format!(
        "ibc-data/{channel}/balance/{}",
        asset.encode_hex::<String>()
    )
}

#[async_trait]
pub(crate) trait StateReadExt: StateRead {
    #[instrument(skip(self))]
    async fn get_account_balance(&self, address: Address, asset: asset::Id) -> Result<u128> {
        let Some(bytes) = self
            .get_raw(&balance_storage_key(address, asset))
            .await
            .context("failed reading raw account balance from state")?
        else {
            debug!("account balance not found, returning 0");
            return Ok(0);
        };
        let Balance(balance) = Balance::try_from_slice(&bytes).context("invalid balance bytes")?;
        Ok(balance)
    }

    #[instrument(skip(self))]
    async fn get_account_nonce(&self, address: Address) -> Result<u32> {
        let bytes = self
            .get_raw(&nonce_storage_key(address))
            .await
            .context("failed reading raw account nonce from state")?;
        let Some(bytes) = bytes else {
            // the account has not yet been initialized; return 0
            return Ok(0);
        };

        let Nonce(nonce) = Nonce::try_from_slice(&bytes).context("invalid nonce bytes")?;
        Ok(nonce)
    }

    #[instrument(skip(self))]
    async fn get_ibc_channel_balance(&self, channel: &ChannelId, asset: asset::Id) -> Result<u128> {
        let Some(bytes) = self
            .get_raw(&channel_balance_storage_key(channel, asset))
            .await
            .context("failed reading ibc channel balance from state")?
        else {
            debug!("ibc channel balance not found, returning 0");
            return Ok(0);
        };
        let Balance(balance) = Balance::try_from_slice(&bytes).context("invalid balance bytes")?;
        Ok(balance)
    }
}

impl<T: StateRead> StateReadExt for T {}

#[async_trait]
pub(crate) trait StateWriteExt: StateWrite {
    #[instrument(skip(self))]
    fn put_account_balance(
        &mut self,
        address: Address,
        asset: asset::Id,
        balance: u128,
    ) -> Result<()> {
        let bytes = Balance(balance)
            .try_to_vec()
            .context("failed to serialize balance")?;
        self.put_raw(balance_storage_key(address, asset), bytes);
        Ok(())
    }

    #[instrument(skip(self))]
    fn put_account_nonce(&mut self, address: Address, nonce: u32) -> Result<()> {
        let bytes = Nonce(nonce)
            .try_to_vec()
            .context("failed to serialize nonce")?;
        self.put_raw(nonce_storage_key(address), bytes);
        Ok(())
    }

    #[instrument(skip(self))]
    fn put_ibc_channel_balance(
        &mut self,
        channel: &ChannelId,
        asset: asset::Id,
        balance: u128,
    ) -> Result<()> {
        let bytes = Balance(balance)
            .try_to_vec()
            .context("failed to serialize balance")?;
        self.put_raw(channel_balance_storage_key(channel, asset), bytes);
        Ok(())
    }
}

impl<T: StateWrite> StateWriteExt for T {}
