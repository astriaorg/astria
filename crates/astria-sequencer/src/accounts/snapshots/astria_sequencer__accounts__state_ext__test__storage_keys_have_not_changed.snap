---
source: crates/astria-sequencer/src/accounts/state_ext.rs
expression: "balance_storage_key(address, id)"
---
accounts/1c0c490f1b5528d8173c5de46d131160e4b2c0c3/balance/000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f
